src/style-spec/expression/definitions/within.js

```diff
@@ -0,0 +1,273 @@
+    // First, check if geometry points of line segments are all inside polygon
+    for (let i = 0; i < line.length; ++i) {
+        if (!pointWithinPolygon(line[i], polygon)) {
+            return false;
+        }
+    }
+
+    // Second, check if there is line segment intersecting polygon edge
+    for (let i = 0; i < line.length - 1; ++i) {
+        if (lineIntersectPolygon(line[i], line[i + 1], polygon)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+function lineStringWithinPolygons(line, polygons) {
+    if (polygons.type === 'Polygon') {
+        return lineStringWithinPolygon(line, polygons.coordinates);
+    }
+    for (let i = 0; i < polygons.coordinates.length; i++) {
+        if (!lineStringWithinPolygon(line, polygons.coordinates[i])) return false;
+    }
+    return true;
+}
+
+function pointsWithinPolygons(ctx: EvaluationContext, polygonGeometry: GeoJSONPolygons, polyBBox: BBox) {
+    const pointBBox = [Infinity, Infinity, -Infinity, -Infinity];
+    const lngLatPoints = [];
+    for (const points of ctx.geometry()) {
+        for (const point of points) {
+            const p = getLngLatPoint(point, ctx.canonicalID());
+            lngLatPoints.push(p);
+            updateBBox(pointBBox, p);
+        }
+    }
+    if (!boxWithinBox(pointBBox, polyBBox)) return false;
+    for (let i = 0; i < lngLatPoints.length; ++i) {
+        if (!pointWithinPolygons(lngLatPoints[i], polygonGeometry)) return false;
+    }
+    return true;
+}
+
+function linesWithinPolygons(ctx: EvaluationContext, polygonGeometry: GeoJSONPolygons, polyBBox: BBox) {
+    const lineBBox = [Infinity, Infinity, -Infinity, -Infinity];
+    const lineCoords = [];
+    for (const line of ctx.geometry()) {
+        const lineCoord = getLngLatPoints(line, ctx.canonicalID());
+        lineCoords.push(lineCoord);
+        calcBBox(lineBBox, lineCoord, 'LineString');
+    }
+    if (!boxWithinBox(lineBBox, polyBBox)) return false;
+    for (let i = 0; i < lineCoords.length; ++i) {
+        if (!lineStringWithinPolygons(lineCoords[i], polygonGeometry)) return false;
+    }
+    return true;
+}
+
+class Within implements Expression {
+    type: Type;
+    geojson: GeoJSON
+    geometries: GeoJSONPolygons;
+    polyBBox: BBox;
+
+    constructor(geojson: GeoJSON, geometries: GeoJSONPolygons) {
+        this.type = BooleanType;
+        this.geojson = geojson;
+        this.geometries = geometries;
+        this.polyBBox = [Infinity, Infinity, -Infinity, -Infinity];
+        calcBBox(this.polyBBox, this.geometries.coordinates, this.geometries.type);
+    }
+
+    static parse(args: $ReadOnlyArray<mixed>, context: ParsingContext) {
+        if (args.length !== 2)
+            return context.error(`'within' expression requires exactly one argument, but found ${args.length - 1} instead.`);
+        if (isValue(args[1])) {
+            const geojson = (args[1]: Object);
+            if (geojson.type === 'FeatureCollection') {
```

Comment:
Can the `within` expression’s sole argument be an expression that evaluates to a geometry, or does it have to be a literal geometry?

Reply:
We decided that the first version of `within` Expression will have inlined GeoJSON Object as arguments. Follow-up would be something like https://github.com/mapbox/mapbox-gl-native-team/issues/181. But it needs to be carefully designed before starting implementation.